prov/lnx/src/lnx_av.c

```diff
@@ -0,0 +1,702 @@
+{
+	int i;
+	struct lnx_address_prov *prov;
+
+	prov = addrs->la_addr_prov;
+	for (i = 0; i < addrs->la_prov_count; i++) {
+		if (!strcmp(prov->lap_prov, "shm"))
+			return prov;
+		prov = next_prov(prov);
+	}
+
+	return NULL;
+}
+
+static int is_local_addr(struct local_prov **shm_prov, struct lnx_addresses *la)
+{
+	int rc;
+	char hostname[FI_NAME_MAX];
+	struct lnx_address_prov *lap_shm;
+
+	/* check the hostname and compare it to mine
+	 * TODO: Is this good enough? or do we need a better way of
+	 * determining if the address is local?
+	 */
+	rc = gethostname(hostname, FI_NAME_MAX);
+	if (rc == -1) {
+		FI_INFO(&lnx_prov, FI_LOG_CORE, "failed to get hostname\n");
+		return -FI_EPERM;
+	}
+
+	lap_shm = lnx_get_peer_shm_addr(la);
+	if (!lap_shm)
+		return -FI_EOPNOTSUPP;
+
+	/* Shared memory address not provided or not local*/
+	if ((lap_shm->lap_addr_count == 0) ||
+		strncasecmp(hostname, la->la_hostname, FI_NAME_MAX))
+		return -FI_EOPNOTSUPP;
+
+	/* badly formed address */
+	if (*shm_prov && (lap_shm->lap_addr_count > 1 ||
+			  lap_shm->lap_addr_count < 0))
+		return -FI_EPROTO;
+
+	return 0;
+}
+
+static void
+lnx_update_msg_entries(struct lnx_qpair *qp,
+		       fi_addr_t (*get_addr)(struct fi_peer_rx_entry *))
+{
+	struct lnx_queue *q = &qp->lqp_unexq;
+	struct lnx_rx_entry *rx_entry;
+	struct dlist_entry *item;
+
+	ofi_spin_lock(&q->lq_qlock);
+	dlist_foreach(&q->lq_queue, item) {
+		rx_entry = (struct lnx_rx_entry *) item;
+		if (rx_entry->rx_entry.addr == FI_ADDR_UNSPEC)
+			rx_entry->rx_entry.addr = get_addr(&rx_entry->rx_entry);
+	}
+	ofi_spin_unlock(&q->lq_qlock);
+}
+
+void
+lnx_foreach_unspec_addr(struct fid_peer_srx *srx,
+			fi_addr_t (*get_addr)(struct fi_peer_rx_entry *))
+{
+	struct lnx_srx_context *ctxt;
+
+	ctxt = (struct lnx_srx_context *) srx->ep_fid.fid.context;
+
+	lnx_update_msg_entries(&ctxt->srx_lep->le_srq.lps_trecv, get_addr);
+	lnx_update_msg_entries(&ctxt->srx_lep->le_srq.lps_recv, get_addr);
+}
+
+static int lnx_peer_map_addrs(struct dlist_entry *prov_table,
+			      struct lnx_peer *lp, struct lnx_addresses *la,
+			      uint64_t flags, void *context)
+{
+	int i, j, rc;
+	struct lnx_peer_prov *lpp;
+	struct lnx_address_prov *lap;
+	struct local_prov_ep *lpe;
+	struct dlist_entry *eps;
+
+	lap = &la->la_addr_prov[0];
+
+	for (i = 0; i < la->la_prov_count; i++) {
+		if (lap->lap_addr_count > LNX_MAX_LOCAL_EPS)
+			return -FI_EPROTO;
+
+		rc = lnx_get_or_create_peer_prov(prov_table, lp, lap->lap_prov,
+						 &lpp);
+		if (rc)
+			return rc;
+
+		lpp->lpp_flags = flags;
+
+		eps = &lpp->lpp_prov->lpv_prov_eps;
+		dlist_foreach_container(eps, struct local_prov_ep, lpe,
+					entry) {
+			struct lnx_local2peer_map *lpm;
+
+			/* if this is a remote peer, don't insert the shm address
+			 * since we will never talk to that peer over shm
+			 */
+			if (!strncasecmp(lpe->lpe_fabric_name, "shm", 3) &&
+				!lp->lp_local)
+				continue;
+
+			lpm = calloc(sizeof(*lpm), 1);
+			if (!lpm)
+				return -FI_ENOMEM;
```

Comment:
If this fails, can it leave an lnx_peer_prov entry in a incomplete state? In other words, should the lpp be freed if this fails?